File: server/server_main/seed.py
"""
Seed script — creates the admin user and default kiosk.
Reads credentials from .env so nothing sensitive is hardcoded.

Usage:
    cd server/server_main
    python seed.py
"""
import os
import secrets
from dotenv import load_dotenv
from app import create_app
from app.extensions import db
from app.models.user import User
from app.models.kiosk import Kiosk

load_dotenv()

ADMIN_NAME     = os.environ.get("ADMIN_NAME",     "EcoCharge Admin")
ADMIN_EMAIL    = os.environ.get("ADMIN_EMAIL",    "admin@ecocharge.ph")
ADMIN_PASSWORD = os.environ.get("ADMIN_PASSWORD", "admin123!")
KIOSK_NAME     = os.environ.get("KIOSK_NAME",     "Kiosk-001")
KIOSK_LOCATION = os.environ.get("KIOSK_LOCATION", "Main Building Lobby")
DEVICE_API_KEY = os.environ.get("DEVICE_API_KEY", secrets.token_hex(32))

if not ADMIN_PASSWORD:
    raise SystemExit("ERROR: ADMIN_PASSWORD is not set in .env — aborting seed.")

app = create_app()

with app.app_context():
    # Admin user
    if not User.query.filter_by(email=ADMIN_EMAIL).first():
        admin = User(
            name=ADMIN_NAME,
            email=ADMIN_EMAIL,
            qr_code=secrets.token_hex(16),
            is_admin=True,
        )
        admin.set_password(ADMIN_PASSWORD)
        db.session.add(admin)
        print(f"Created admin: {ADMIN_EMAIL}")
    else:
        print(f"Admin already exists: {ADMIN_EMAIL} (skipped)")

    # Default kiosk
    if not Kiosk.query.filter_by(name=KIOSK_NAME).first():
        kiosk = Kiosk(
            name=KIOSK_NAME,
            location=KIOSK_LOCATION,
            api_key=DEVICE_API_KEY,
        )
        db.session.add(kiosk)
        db.session.commit()
        kiosk = Kiosk.query.filter_by(name=KIOSK_NAME).first()
        print(f"Created kiosk: {KIOSK_NAME}")
        print(f"  Kiosk API key: {kiosk.api_key}")
        print(f"  >> Set DEVICE_API_KEY={kiosk.api_key} in your .env and ESP32 firmware")
    else:
        kiosk = Kiosk.query.filter_by(name=KIOSK_NAME).first()
        print(f"Kiosk already exists: {KIOSK_NAME} (skipped)")
        print(f"  Kiosk API key: {kiosk.api_key}")

    db.session.commit()
    print("Seed complete.")

File: server/server_main/prisma/seed.ts
/**
 * Manual seed — run with:  npm run seed
 *
 * Wipes ALL data in FK-safe order, then creates:
 *   1. Admin user       (ADMIN_EMAIL / ADMIN_PASSWORD / ADMIN_NAME)
 *   2. Default kiosk    (KIOSK_NAME / KIOSK_LOCATION / DEVICE_API_KEY)
 *   3. System settings  (hardcoded defaults)
 *
 * Never runs automatically on boot. To run on Render: Shell tab → npm run seed
 */

import "dotenv/config";
import crypto from "crypto";
import bcrypt from "bcryptjs";
import { PrismaClient } from "@prisma/client";
import { SETTING_DEFAULTS } from "../src/services/settingsService";

const prisma = new PrismaClient();

async function main() {
  console.log("\n════════════════════════════════════════");
  console.log("  EcoCharge — Full DB Reset + Seed");
  console.log("════════════════════════════════════════\n");

  // ── 1. Wipe all data (leaf tables first to satisfy FK constraints) ──────────
  console.log("Wiping all data…");
  await prisma.deviceCommand.deleteMany();
  await prisma.deviceTelemetry.deleteMany();
  await prisma.chargingSession.deleteMany();
  await prisma.creditTransaction.deleteMany();
  await prisma.bottleDeposit.deleteMany();
  await prisma.kioskSession.deleteMany();
  await prisma.kiosk.deleteMany();
  await prisma.user.deleteMany();
  await prisma.systemSetting.deleteMany();
  console.log("  ✔  All tables cleared\n");

  // ── 2. Admin user ───────────────────────────────────────────────────────────
  const email = process.env.ADMIN_EMAIL ?? "admin@ecocharge.ph";
  const name = process.env.ADMIN_NAME ?? "EcoCharge Admin";
  const password = process.env.ADMIN_PASSWORD;

  if (!password) {
    console.error("  ✖  ADMIN_PASSWORD not set — set it in .env and retry");
    process.exit(1);
  }

  const admin = await prisma.user.create({
    data: {
      name,
      email,
      passwordHash: await bcrypt.hash(password, 12),
      qrCode: crypto.randomBytes(16).toString("hex"),
      isAdmin: true,
    },
  });
  console.log(`  ✔  Admin   — #${admin.id}  ${admin.email}`);

  // ── 3. Default kiosk ────────────────────────────────────────────────────────
  const apiKey =
    process.env.DEVICE_API_KEY ??
    "SET_AT_BUILD_TIME";
  const kioskName = process.env.KIOSK_NAME ?? "Kiosk-001";
  const kioskLocation =
    process.env.KIOSK_LOCATION ?? "UC Lapu-Lapu and Mandaue";

  const kiosk = await prisma.kiosk.create({
    data: { name: kioskName, location: kioskLocation, apiKey },
  });
  console.log(`  ✔  Kiosk   — #${kiosk.id}  "${kiosk.name}"`);
  console.log(`       apiKey = ${kiosk.apiKey}`);
  console.log(
    `       → set NEXT_PUBLIC_KIOSK_ID=${kiosk.id} on kiosk web Render service`,
  );

  // ── 4. System settings ──────────────────────────────────────────────────────
  await prisma.systemSetting.createMany({
    data: Object.entries(SETTING_DEFAULTS).map(([key, value]) => ({
      key,
      value,
    })),
  });
  console.log(
    `  ✔  Settings — ${Object.keys(SETTING_DEFAULTS).length} keys seeded`,
  );

  // ── Summary ─────────────────────────────────────────────────────────────────
  console.log("\n════════════════════════════════════════");
  console.log("  Seed complete ✔");
  console.log("════════════════════════════════════════\n");
}

main()
  .catch((err) => {
    console.error("\n  ✖  Seed failed:", err);
    process.exit(1);
  })
  .finally(() => prisma.$disconnect());
